### PR TITLE
EZP-27352: App should handle content discover event

### PR DESCRIPTION
**Pull request by StephaneDiot**
jira: https://jira.ez.no/browse/EZP-27352

## Description 

The app component should be responsible of handling the content discover event (event that launch UDW).

Goes with https://github.com/ezsystems/PlatformUIBundle/pull/876 and https://github.com/ezsystems/hybrid-platform-ui/pull/16

## tests

unit tested

### CODE REVIEW

File: js/ez-platform-ui-app.js [hunk 1]
```diff
@@ -123,6 +123,32 @@
         constructor() {
             super();
             this._enhanceNavigation();
+            this._handleContentDiscover();
+        }
+
+        /**
+         * Adds event listener for 'contentDiscover' event.
```
Comment (dpobel) [inline js/ez-platform-ui-app.js]: `ez:contentDiscover` actually (with backticks instead of simple quotes)

File: js/ez-platform-ui-app.js [hunk 2]
```diff
@@ -123,6 +123,32 @@
...
+         * The event listener is responsible for creating <ez-universal-discovery> custom element.
+         *
```
Comment (dpobel) [inline js/ez-platform-ui-app.js]: CS

File: js/ez-platform-ui-app.js [hunk 3]
```diff
@@ -123,6 +123,32 @@
...
+         */
+        _handleContentDiscover() {
+            this.addEventListener('ez:contentDiscover', function (e) {
+                const {config, handlers} = e.detail;
```
Comment (dpobel) [inline js/ez-platform-ui-app.js]: this is a quite long and non trivial function, it deserves to be a named method and to have a short doc to explain what it does

File: js/ez-platform-ui-app.js [hunk 4]
```diff
@@ -123,6 +123,32 @@
...
+
+                Object.keys(config).forEach(function (prop) {
+                    udw[prop] = config[prop];
+                });
+                Object.keys(handlers).forEach(function (event) {
+                    udw.addEventListener(event, handlers[event]);
+                });
+                udw.addEventListener('ez:confirm', removeUdw);
+                udw.addEventListener('ez:cancel', removeUdw);
+                document.body.appendChild(udw);
```
Comment (dpobel) [inline js/ez-platform-ui-app.js]: since the app is now responsible for creating the `<ez-universal-discovery>` element, it should be created inside the app not as a sibling element under body (and that's also why the `ez-js-standard-form` is needed on the `<ez-universal-discovery>`, while if it was created directly under the body, this class would not be necessary)

File: test/js/ez-platform-ui-app.js [hunk 5]
```diff
@@ -150,6 +150,78 @@ describe('ez-platform-ui-app', function() {
...
+                        anyEventName: () => {
+                            eventFired = true;
+                        },
+                    },
+                },
+            }));
+        }
+
+        afterEach(function () {
+            if (document.querySelector('ez-universal-discovery')) {
```
Comment (dpobel) [inline test/js/ez-platform-ui-app.js]: please use a local variable to avoid calling `document.querySelector('ez-universal-discovery')` 3 times in 2 lines

File: test/js/ez-platform-ui-app.js [hunk 6]
```diff
@@ -150,6 +150,78 @@ describe('ez-platform-ui-app', function() {
         });
     });
 
+    describe('handle contentDiscover', function () {
```
Comment (dpobel) [inline test/js/ez-platform-ui-app.js]:
```js
describe('handle `ez:contentDiscover`', function () {
```

**Review by dpobel — CHANGES_REQUESTED** [review]

File: test/js/ez-platform-ui-app.js [hunk 7]
```diff
@@ -150,6 +150,78 @@ describe('ez-platform-ui-app', function() {
...
+        const property = 'anything';
+        let eventFired = false;
+
+        function fireContentDiscover () {
+            element.dispatchEvent(new CustomEvent('ez:contentDiscover', {
+                detail: {
+                    config: {
+                        anyProp: property,
+                    },
+                    handlers: {
```
Comment (bdunogier) [inline test/js/ez-platform-ui-app.js:163]: Why `handlers` ? Aren't they used as `listeners` ?

File: test/js/ez-platform-ui-app.js [hunk 8]
```diff
@@ -150,6 +150,78 @@ describe('ez-platform-ui-app', function() {
...
+            const udw = document.querySelector('ez-universal-discovery');
+
+            assert.equal(
+                udw.anyProp,
+                property,
+                'universal discovery component should get the properties given in the event'
+            );
+        });
+
+        it('should set the handlers to the universal discovery widget', function () {
```
Comment (bdunogier) [inline test/js/ez-platform-ui-app.js]: I'd rephrase this a bit (setting XXX to the UDW doesn't sound right). What about `should register the provided handlers as universal discovery widget listeners` ? It describes what is done better I think.

Sorry if I'm off-topic :)

**Review by bdunogier — COMMENTED** [review]
I know that there are historical reasons why you name those things handlers, but it's still a very very generic name :)

File: test/js/ez-platform-ui-app.js (hunk 7 above)
Comment (StephaneDiot) [reply test/js/ez-platform-ui-app.js:163]: actually here it's the struct of the detail attr I get from the event facade. To change this struct I need to alter some code out of this repo I guess...

Comment (dpobel) [reply test/js/ez-platform-ui-app.js:163]: @StephaneDiot  Bertrand is right and it's the right time to change that

File: js/ez-platform-ui-app.js [hunk 9]
```diff
@@ -123,6 +123,37 @@
...
             super();
             this._enhanceNavigation();
+            this._handleContentDiscover();
+        }
+
+        /**
+         * Adds event listener for `ez:contentDiscover` event.
+         */
+        _handleContentDiscover() {
+            this.addEventListener('ez:contentDiscover', this._createUDCustomElement);
```
Comment (dpobel) [inline js/ez-platform-ui-app.js]: I think you need to bind this method to `this`

**Review by dpobel — CHANGES_REQUESTED** [review]

File: test/js/ez-platform-ui-app.js (hunk 7 above)
Comment (dpobel) [reply test/js/ez-platform-ui-app.js:163]: note that at the moment, only `ez-browse` uses the UD and you are working on it :)

Comment (StephaneDiot) [reply test/js/ez-platform-ui-app.js:163]: I agree with the change but  shouldn't this  be done with the implementation of ez-browse custom element then ?

Comment (dpobel) [reply test/js/ez-platform-ui-app.js:163]: as long as it's not forgotten, that works for me

**Review by dpobel — APPROVED** [review]
